### PR TITLE
fix QuantizeAvx512

### PR DESCRIPTION
**Pull request by zhuhaozhe**
For quantize
```
  for (; i < len / VLEN * VLEN; i += VLEN) {
    __m512 x_vals = _mm512_load_ps(src + i);
    __m512 x_transformed_v = _mm512_mul_ps(x_vals, inverse_scale_v);
    x_transformed_v =
        _mm512_min_ps(x_transformed_v, _mm512_set1_ps(int32_float_max_val));
    __m512i x_rounded_v = _mm512_cvtps_epi32(x_transformed_v);
    x_rounded_v = _mm512_add_epi32(x_rounded_v, _mm512_set1_epi32(zero_point));
    __m512i x_clipped_v =
        _mm512_max_epi32(min_v, _mm512_min_epi32(max_v, x_rounded_v));

    x_clipped_v = _mm512_shuffle_epi8(x_clipped_v, shuffle_mask_v);
    x_clipped_v = _mm512_permutexvar_epi32(permute_mask_l8_v, x_clipped_v);
    _mm_storeu_si128(
        reinterpret_cast<__m128i*>(dst + i),
        _mm512_castsi512_si128(x_clipped_v));
  }
```

```
    x_clipped_v = _mm512_shuffle_epi8(x_clipped_v, shuffle_mask_v);
    x_clipped_v = _mm512_permutexvar_epi32(permute_mask_l8_v, x_clipped_v);
```
is aiming to cast `int32` to `int8` and shuffle 16 `int8` to the first 128 bits.

For example, `A1` represent 8bit
```
    x_clipped_v = _mm512_shuffle_epi8(x_clipped_v, shuffle_mask_v);
    A1A2A3**A4** B1B2B3**B4** C1C2C3**C4** D1D2D3**D4**            -> D4C4B4A4 other 32 * 3 bit
    E1E2E3**E4** F1F2F3**F4** G1G2G3**G4** H1H2H3**H4**            -> H4G4F4E4 other 32 * 3 bit
    I1I2I3**I4** J1J2J3**J4** K1K2K3**K4** L1L2L3**L4**            -> L4K4J4I4 other 32 * 3 bit
    M1M2M3**M4** N1N2N3**N4** O1O2O3**O4** P1P2P3**P4**            -> P4O4N4M4 other 32 * 3 bit
    x_clipped_v = _mm512_permutexvar_epi32(permute_mask_l8_v, x_clipped_v);
    D4C4B4A4 other 32 * 3 bit        -> D4C4B4A4 H4G4F4E4 L4K4J4I4 P4O4N4M4 
    H4G4F4E4 other 32 * 3 bit           other 3 * 4 * 32 bits
    L4K4J4I4 other 32 * 3 bit
    P4O4N4M4 other 32 * 3 bit

```

Based on https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm512_permutexvar_epi32&ig_expand=4966,5088.
```
FOR j := 0 to 15
	i := j*32
	id := idx[i+3:i]*32
	dst[i+31:i] := a[id+31:id]
ENDFOR
dst[MAX:512] := 0
```
the `permute_mask_l8_v` should satisfy
```
permute_mask_l8_v[3:0] = 0
permute_mask_l8_v[3 + 32:0 + 32] = 4
permute_mask_l8_v[3 + 64:0 + 64] = 8
permute_mask_l8_v[3 + 96:0 + 96] = 12
```
The other part of `permute_mask_l8_v` does not matters.

`AVX2` version is correct.

It is not exposed before it is only called with fixed length `64` https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/vec512/vec512_qint.h#L545-L546.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104400

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — CHANGES_REQUESTED** [review]
Can you add test to cover this fix?

**Comment by zhuhaozhe** [comment]
> Can you add test to cover this fix?

This `QuantizeAVX512` or `QuantizeAVX2` are not directly used by kernels except `qavg_pool2d_nhwc`.
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L1721-L1735

This `avg_pool2d_nhwc` seems are not well-tested and should have other issues:
https://github.com/pytorch/pytorch/blob/main/test/quantization/core/test_quantized_op.py#L1678-L1681
`QuantizeAvx512` and `QuantizeAvx2` should not support qint32 but it will also call them.

And Pytorch now seems always dispatch to `AVX2`.

So I only added `QuantizeAvx512` and `QuantizeAvx2` test in `vec_test_all_types.cpp` at `VEC` level.

File: aten/src/ATen/test/vec_test_all_types.cpp [hunk 1]
```diff
@@ -1135,6 +1146,75 @@ namespace {
             if (AssertVectorized<vec>(NAME_INFO(Quantize), expected, actual).check()) return;
         } //trials;
     }
+#if !defined(CPU_CAPABILITY_DEFAULT) && !defined(_MSC_VER)
+    // This testcase aim to test at::vec::QuantizeAvx512 and
```
Comment (jgong5) [inline aten/src/ATen/test/vec_test_all_types.cpp:1150]:
```suggestion
    // This test case aims to test at::vec::QuantizeAvx512 and
```

File: aten/src/ATen/test/vec_test_all_types.cpp [hunk 2]
```diff
@@ -1135,6 +1146,75 @@ namespace {
             if (AssertVectorized<vec>(NAME_INFO(Quantize), expected, actual).check()) return;
         } //trials;
     }
+#if !defined(CPU_CAPABILITY_DEFAULT) && !defined(_MSC_VER)
```
Comment (jgong5) [inline aten/src/ATen/test/vec_test_all_types.cpp:1149]: Suggest to make it specific to avx256 and avx516 here.
```suggestion
#if (defined(CPU_CAPABILITY_AVX2) ||  defined(CPU_CAPABILITY_AVX512)) && !defined(_MSC_VER)
```

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 3]
```diff
@@ -1719,14 +1719,14 @@ void do_avg_pool_nhwc_on_AVX_n(
       // first quantize using AVX2 or AVX512 using 32 lanes, then 8, finally falls
       // back to single
 #ifdef CPU_CAPABILITY_AVX2
-      QuantizeAvx2<T>(
+      QuantizeAvx2<typename T::underlying>(
```
Comment (jgong5) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:1722]: May I know why this change is necessary?

Comment (zhuhaozhe) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:1722]: The initial reason for doing this is I failed to deduce `c10:qint8` or `c10:quint8` while invoking `QuantizeAvx2/QuantizeAvx512` in the test. And I found it is non-necessary to know about `c10:qint8` or `c10:quint8` in `QuantizeAvx2/QuantizeAvx512`. What need to know is only the `underlying`. I also found `RequantizeAvx512/RequantizeAvx2` is using the `underlying` directly. So I do this change.

File: aten/src/ATen/test/vec_test_all_types.cpp (hunk 2 above)
Comment (zhuhaozhe) [reply aten/src/ATen/test/vec_test_all_types.cpp]: Thanks for advice, done.

File: aten/src/ATen/test/vec_test_all_types.cpp (hunk 1 above)
Comment (zhuhaozhe) [reply aten/src/ATen/test/vec_test_all_types.cpp]: Thanks for the advice, done.

**Review by jgong5 — APPROVED** [review]

**Review by mingfeima — APPROVED** [review]
good catch!

**Comment by zhuhaozhe** [comment]
@jerryzh168, I can't merge it. May you help to review this PR?

**Review by jerryzh168 — APPROVED** [review]

**Comment by jerryzh168** [comment]
> @jerryzh168, I can't merge it. May you help to review this PR?

please try again

**Comment by zhuhaozhe** [comment]
> > @jerryzh168, I can't merge it. May you help to review this PR?
> 
> please try again

Thanks, it can be merged now.
